dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -311,18 +311,34 @@ ExchangeReceiverBase<RPCContext>::ExchangeReceiverBase(
 template <typename RPCContext>
 ExchangeReceiverBase<RPCContext>::~ExchangeReceiverBase()
 {
-    setState(ExchangeReceiverState::CLOSED);
-    msg_channel.finish();
+    close();
     thread_manager->wait();
 }
 
 template <typename RPCContext>
 void ExchangeReceiverBase<RPCContext>::cancel()
 {
+    auto cur_state = getState();
+    if (cur_state == ExchangeReceiverState::CANCELED || cur_state == ExchangeReceiverState::CLOSED)
+    {
+        return;
+    }
     setState(ExchangeReceiverState::CANCELED);
     msg_channel.finish();
 }
 
+template <typename RPCContext>
+void ExchangeReceiverBase<RPCContext>::close()
+{
+    auto cur_state = getState();
+    if (cur_state == ExchangeReceiverState::CANCELED || cur_state == ExchangeReceiverState::CLOSED)
+    {
+        return;
+    }
+    setState(ExchangeReceiverState::CLOSED);
+    msg_channel.finish();
+}
```

Comment:
it seems that those lines are almost similar to L321-327

Reply:
refine in another pr, since my local repo is broken, see another PR https://github.com/pingcap/tiflash/pull/4441